Stop stripping prefixes from PseudoClass & PseudoElement

nt1m opened the pull request:
#### 55c55cb3d5db644cbfa46baed5b80b88cafd73cf
<pre>
Stop stripping prefixes from PseudoClass &amp; PseudoElement
<a href="https://bugs.webkit.org/show_bug.cgi?id=266946">https://bugs.webkit.org/show_bug.cgi?id=266946</a>
<a href="https://rdar.apple.com/120323601">rdar://120323601</a>

Reviewed by Darin Adler.

Since prefixes are no longer the main practice to graduate experimental features and they are only used for compatibility reasons or for internal usage,
there is no longer a reason to strip the prefixes from enum classes.

It provides more clarity to what is exposed to the web, and it reduces confusion.

This also fixes the bug where the WebInspector shows unprefixed pseudo elements for scrollbar/resizer pseudo elements which don&apos;t actually have unprefixed counterparts.

* LayoutTests/inspector/css/getMatchedStylesForNode-expected.txt:
* Source/JavaScriptCore/inspector/scripts/codegen/generator.py:
(ucfirst):
* Source/JavaScriptCore/inspector/protocol/CSS.json:
* Source/WebCore/animation/WebAnimationUtilities.cpp:
(WebCore::compareDeclarativeAnimationOwningElementPositionsInDocumentTreeOrder):
(WebCore::pseudoIdAsString):
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::pseudoId):
(WebCore::CSSSelector::selectorText const):
* Source/WebCore/css/CSSSelector.h:
(WebCore::isLogicalCombinationPseudoClass):
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::hasScrollbarPseudoElement):
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/SelectorChecker.h:
(WebCore::SelectorChecker::isCommonPseudoClassSelector):
* Source/WebCore/css/SelectorPseudoClassAndCompatibilityElementMap.in:
* Source/WebCore/css/makeSelectorPseudoClassAndCompatibilityElementMap.py:
(enumerablePseudoType):
* Source/WebCore/css/makeSelectorPseudoElementsMap.py:
(enumerablePseudoType):
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::isPseudoClassValidAfterPseudoElement):
(WebCore::isOnlyPseudoClassFunction):
(WebCore::CSSSelectorParser::consumePseudo):
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType):
(WebCore::SelectorCompiler::constructFragmentsInternal):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::setBeingDragged):
(WebCore::Element::setIsLink):
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::setAnimatingFullscreen):
(WebCore::FullscreenManager::setFullscreenControlsHidden):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::setAutoFilledAndViewable):
(WebCore::HTMLInputElement::setAutoFilledAndObscured):
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::protocolValueForPseudoId):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::styleHidesScrollbarWithOrientation const):
(WebCore::LocalFrameView::updateScrollCorner):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::updateScrollCornerStyle):
(WebCore::RenderLayerScrollableArea::updateResizerStyle):
* Source/WebCore/rendering/RenderScrollbar.cpp:
(WebCore::pseudoForScrollbarPart):
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::usesLegacyScrollbarStyle const):
* Source/WebCore/style/RuleSet.cpp:
(WebCore::Style::RuleSet::addRule):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolveElement):
(WebCore::Style::TreeResolver::resolvePseudoElement):
* Source/WebInspectorUI/UserInterface/Controllers/CSSManager.js:
(WI.CSSManager.displayNameForPseudoId):

Canonical link: <a href="https://commits.webkit.org/272544@main">https://commits.webkit.org/272544@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/992d809ded76e58c79ebd054d1f62231cfafb5af

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32037 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10751 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/33818 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/34575 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29031 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32862 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13109 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7978 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/28619 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32431 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/9079 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/28637 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7871 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/8046 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/28552 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35919 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/27522 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/29141 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/29007 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/34151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/32118 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8151 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/6113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/32009 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/9791 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/38547 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8798 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/8182 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4164 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8673 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/CSSSelector.h:
```diff
@@ -116,25 +116,25 @@ struct PossiblyQuotedIdentifier {
             Default,
             Disabled,
-            HtmlDocument, // for internal use only with forms in table case in UA stylesheet
+            InternalHtmlDocument, // for internal use in html.css
```
Given we are spelling `-webkit-` as `WebKit`, could we make this `InternalHTMLDocument`?

nt1m replied:
Sure, I special cased it

darinadler commented on Source/WebCore/css/SelectorChecker.cpp:
```diff
@@ -222,13 +222,20 @@ bool SelectorChecker::matchHostPseudoClass(const CSSSelector& selector, const El
 inline static bool hasScrollbarPseudoElement(const PseudoIdSet& dynamicPseudoIdSet)
 {
-    PseudoIdSet scrollbarIdSet = { PseudoId::Scrollbar, PseudoId::ScrollbarThumb, PseudoId::ScrollbarButton, PseudoId::ScrollbarTrack, PseudoId::ScrollbarTrackPiece, PseudoId::ScrollbarCorner };
+    PseudoIdSet scrollbarIdSet = {
```
Can we use `constexpr` here?

nt1m replied:
It's actually not trivial here. The class isn't designed for it

darinadler commented on Source/WebCore/css/makeSelectorPseudoClassAndCompatibilityElementMap.py:
```diff
@@ -31,20 +31,14 @@
     webkitPrefix = '-webkit-'
     if (stringPseudoType.startswith(webkitPrefix)):
-        stringPseudoType = stringPseudoType[len(webkitPrefix):]
+        stringPseudoType = stringPseudoType.replace('-webkit-', '-WebKit-')
```
I’m not sure we should keep a local variable `webkitPrefix` if we only use it once.

darinadler commented on Source/WebCore/css/makeSelectorPseudoElementsMap.py:
```diff
@@ -31,9 +31,14 @@
     webkitPrefix = '-webkit-'
     if (stringPseudoType.startswith(webkitPrefix)):
-        stringPseudoType = stringPseudoType[len(webkitPrefix):]
+        stringPseudoType = stringPseudoType.replace('-webkit-', '-WebKit-')
```
Same here. I’m not sure we should keep a local variable `webkitPrefix` if we only use it once.

darinadler commented on Source/WebCore/rendering/style/RenderStyleConstants.h:
```diff
@@ -103,17 +103,17 @@ enum class PseudoId : uint32_t {
-    ScrollbarTrackPiece,
-    ScrollbarCorner,
-    Resizer,
+    WebKitScrollbarThumb,
```
If these are really “internal” then maybe we should use a different kind of prefix?

nt1m replied:
These are web exposed, so they're not really internal in the sense that `-internal-` is used for. I'll investigate this as a follow up to disambiguate the meaning of "Internal" here.

nt1m committed:
Stop stripping prefixes from PseudoClass & PseudoElement
https://bugs.webkit.org/show_bug.cgi?id=266946
rdar://120323601

Reviewed by Darin Adler.

Since prefixes are no longer the main practice to graduate experimental features and they are only used for compatibility reasons or for internal usage,
there is no longer a reason to strip the prefixes from enum classes.

It provides more clarity to what is exposed to the web, and it reduces confusion.

This also fixes the bug where the WebInspector shows unprefixed pseudo elements for scrollbar/resizer pseudo elements which don't actually have unprefixed counterparts.

* LayoutTests/inspector/css/getMatchedStylesForNode-expected.txt:
* Source/JavaScriptCore/inspector/scripts/codegen/generator.py:
(ucfirst):
* Source/JavaScriptCore/inspector/protocol/CSS.json:
* Source/WebCore/animation/WebAnimationUtilities.cpp:
(WebCore::compareDeclarativeAnimationOwningElementPositionsInDocumentTreeOrder):
(WebCore::pseudoIdAsString):
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::pseudoId):
(WebCore::CSSSelector::selectorText const):
* Source/WebCore/css/CSSSelector.h:
(WebCore::isLogicalCombinationPseudoClass):
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::hasScrollbarPseudoElement):
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/SelectorChecker.h:
(WebCore::SelectorChecker::isCommonPseudoClassSelector):
* Source/WebCore/css/SelectorPseudoClassAndCompatibilityElementMap.in:
* Source/WebCore/css/makeSelectorPseudoClassAndCompatibilityElementMap.py:
(enumerablePseudoType):
* Source/WebCore/css/makeSelectorPseudoElementsMap.py:
(enumerablePseudoType):
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::isPseudoClassValidAfterPseudoElement):
(WebCore::isOnlyPseudoClassFunction):
(WebCore::CSSSelectorParser::consumePseudo):
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType):
(WebCore::SelectorCompiler::constructFragmentsInternal):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::setBeingDragged):
(WebCore::Element::setIsLink):
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::setAnimatingFullscreen):
(WebCore::FullscreenManager::setFullscreenControlsHidden):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::setAutoFilledAndViewable):
(WebCore::HTMLInputElement::setAutoFilledAndObscured):
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::protocolValueForPseudoId):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::styleHidesScrollbarWithOrientation const):
(WebCore::LocalFrameView::updateScrollCorner):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::updateScrollCornerStyle):
(WebCore::RenderLayerScrollableArea::updateResizerStyle):
* Source/WebCore/rendering/RenderScrollbar.cpp:
(WebCore::pseudoForScrollbarPart):
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator<<):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::usesLegacyScrollbarStyle const):
* Source/WebCore/style/RuleSet.cpp:
(WebCore::Style::RuleSet::addRule):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolveElement):
(WebCore::Style::TreeResolver::resolvePseudoElement):
* Source/WebInspectorUI/UserInterface/Controllers/CSSManager.js:
(WI.CSSManager.displayNameForPseudoId):

Canonical link: https://commits.webkit.org/272544@main
